### PR TITLE
Adds AVX2 variants of Kyber512, Kyber768, Kyber1024

### PR DESCRIPTION
**Pull request by thomwiggers**
Adds the AVX2 variants of the Kyber packages.

This branch requires #224

#### Manually checked properties

* [x] No stringification macros
* [ ] Output-parameter pointers in functions are on the left
* [ ] Negative return values on failure of API functions (within restrictions of FO transform).
* [x] `const` arguments are labeled as `const`
* [x] variable declarations at the beginning (except in `for (size_t i=...`)
* Optional:
  * [ ] All integer types are of fixed size, using `stdint.h` types (
    * [x] including `uint8_t` instead of `unsigned char`
  * [x] Integers used for indexing are of size `size_t`

### CODE REVIEW

**Comment by thomwiggers** [comment]
I have been able to solve a lot of problems, and the C files now work with Microsoft's CL compiler. 

However, it also turns out Microsoft's assembler (masm, `ml.exe`) uses a different syntax – and I am not sure how to support both GNU `as` and `ml.exe` from the same source files...

**Comment by thomwiggers** [comment]
I changed #224 to also allow specifying OS limits. The AVX2 folders will be skipped on Windows.

Meanwhile; I should fix the cast problems in Kyber-clean.

**Comment by thomwiggers** [comment]
Ok, that should finish up all that needs to be done for this PR to be merged.

File: crypto_kem/kyber1024/avx2/shuffle.s [hunk 1]
```diff
@@ -0,0 +1,252 @@
+.include "fq.inc"
+.include "shuffle.inc"
+
+/*
+.global nttpack_avx
```
Comment (mkannwischer) [inline crypto_kem/kyber1024/avx2/shuffle.s]: remove?

File: crypto_kem/kyber1024/clean/poly.c [hunk 2]
```diff
@@ -27,100 +25,100 @@ void PQCLEAN_KYBER1024_CLEAN_poly_compress(unsigned char *r, poly *a) {
...
+*              - const uint8_t *a: pointer to input byte array (of KYBER_POLYBYTES bytes)
 **************************************************/
-void PQCLEAN_KYBER1024_CLEAN_poly_frombytes(poly *r, const unsigned char *a) {
+void PQCLEAN_KYBER1024_CLEAN_poly_frombytes(poly *r, const uint8_t *a) {
     int i;
 
     for (i = 0; i < KYBER_N / 2; i++) {
-        r->coeffs[2 * i]   = a[3 * i]        | ((uint16_t)a[3 * i + 1] & 0x0f) << 8;
-        r->coeffs[2 * i + 1] = a[3 * i + 1] >> 4 | ((uint16_t)a[3 * i + 2] & 0xff) << 4;
+        r->coeffs[2 * i]   = (int16_t)(a[3 * i]        | ((uint16_t)a[3 * i + 1] & 0x0f) << 8);
```
Comment (mkannwischer) [inline crypto_kem/kyber1024/clean/poly.c]: fix alignment/spacing?

File: crypto_kem/kyber1024/clean/polyvec.c [hunk 3]
```diff
@@ -24,81 +23,81 @@ void PQCLEAN_KYBER1024_CLEAN_polyvec_compress(unsigned char *r, polyvec *a) {
...
-            r[11 * j + 2] = ((t[1] >>  5) | ((t[2] & 0x03) << 6)) & 0xff;
-            r[11 * j + 3] =  (t[2] >>  2) & 0xff;
-            r[11 * j + 4] = ((t[2] >> 10) | ((t[3] & 0x7f) << 1)) & 0xff;
-            r[11 * j + 5] = ((t[3] >>  7) | ((t[4] & 0x0f) << 4)) & 0xff;
-            r[11 * j + 6] = ((t[4] >>  4) | ((t[5] & 0x01) << 7)) & 0xff;
-            r[11 * j + 7] =  (t[5] >>  1) & 0xff;
-            r[11 * j + 8] = ((t[5] >>  9) | ((t[6] & 0x3f) << 2)) & 0xff;
-            r[11 * j + 9] = ((t[6] >>  6) | ((t[7] & 0x07) << 5)) & 0xff;
-            r[11 * j + 10] = (t[7] >>  3) & 0xff;
+            r[11 * j + 0] = (uint8_t)( t[0] & 0xff);
```
Comment (mkannwischer) [inline crypto_kem/kyber1024/clean/polyvec.c]: Can't you remove all the & 0xff if you explicitly cast to uint8?

File: crypto_kem/kyber1024/clean/verify.c [hunk 4]
```diff
@@ -36,12 +35,12 @@ unsigned char PQCLEAN_KYBER1024_CLEAN_verify(const unsigned char *a, const unsig
...
-* Arguments:   unsigned char *r:       pointer to output byte array
-*              const unsigned char *x: pointer to input byte array
+* Arguments:   uint8_t *r:       pointer to output byte array
+*              const uint8_t *x: pointer to input byte array
 *              size_t len:             Amount of bytes to be copied
-*              unsigned char b:        Condition bit; has to be in {0,1}
+*              uint8_t b:        Condition bit; has to be in {0,1}
 **************************************************/
-void PQCLEAN_KYBER1024_CLEAN_cmov(unsigned char *r, const unsigned char *x, size_t len, unsigned char b) {
+void PQCLEAN_KYBER1024_CLEAN_cmov(uint8_t *r, const unsigned char *x, size_t len, unsigned char b) {
```
Comment (mkannwischer) [inline crypto_kem/kyber1024/clean/verify.c]: Do you want to replace all the unsigned char? Here it's inconsistent.

Comment (thomwiggers) [reply crypto_kem/kyber1024/clean/verify.c]: Need to grep a little bit harder still :p

**Pull request merged**